src/MongoDB.Driver/TransactionExecutor.cs

```diff
@@ -0,0 +1,308 @@
+                return new WithTransactionCallbackOutcome<TResult>(result, false);
+            }
+            catch (Exception ex)
+            {
+                if (clientSession.IsInTransaction)
+                {
+                    await clientSession.AbortTransactionAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                if (ex is MongoException mongoException)
+                {
+                    var now = dateTimeProvider.UtcNow;
+                    if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !HasTimedOut(startTime, now))
+                    {
+                        return new WithTransactionCallbackOutcome<TResult>(default, true);
+                    }
+
+                    if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel))
+                    {
+                        await clientSession.CommitTransactionAsync(cancellationToken).ConfigureAwait(false);
+                    }
+                }
+
+                throw;
+            }
+        }
+
+        private static bool WithTransactionCommitProcessingWithRetry(IClientSessionHandle clientSession, DateTime startTime, IDateTimeProvider dateTimeProvider, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    clientSession.CommitTransaction(cancellationToken);
+                    return true;
+                }
+                catch (Exception ex)
+                {
+                    if (ex is MongoException mongoException)
+                    {
+                        var now = dateTimeProvider.UtcNow;
+                        if (mongoException.HasErrorLabel(UnknownTransactionCommitResultLabel) &&
+                            !IsWriteConcernTimeoutError(mongoException) &&
+                            !HasTimedOut(startTime, now))
+                        {
+                            continue;
+                        }
+
+                        if (mongoException.HasErrorLabel(TransientTransactionErrorLabel) && !HasTimedOut(startTime, now))
+                        {
+                            // the transaction will be full processed once again.
+                            return false;
+                        }
+                    }
+
+                    throw;
+                }
+            }
+        }
+
+        private static async Task<bool> WithTransactionCommitProcessingWithRetryAsync(IClientSessionHandle clientSession, DateTime startTime, IDateTimeProvider dateTimeProvider, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+
+                    await clientSession.CommitTransactionAsync(cancellationToken).ConfigureAwait(false);
+                    return false;
```

Comment:
return true?

Why did no test fail?

Reply:
I didn't rename/change processing of `CommitWithRetries` for sync and async processing properly when I moved logic into TransactionExecuter.
I've fixed it.